src/gpgmm/utils/ContiguousLinkNodeAllocator.h

```diff
@@ -0,0 +1,113 @@
+      public:
+        explicit ContiguousLinkNodeAllocator(uint64_t objectsPerBlock)
+            : mObjectsPerBlock(objectsPerBlock) {
+        }
+
+        ~ContiguousLinkNodeAllocator() {
+            for (std::vector<ContiguousLinkNodeAllocation>* blockPtr : mAllocations) {
+                delete blockPtr;
+            }
+        }
+
+        using AllocationAddress = std::pair<uint64_t, uint64_t>;
+
+        // Using this struct we can reinterpret_cast<T*> when we need to provide the allocation
+        // pointer because it is the first object in the struct. We can
+        // reinterpret_cast<ContiguousLinkNodeAllocation> back from the allocation pointer to get
+        // the AllocationAddress indices needed during Deallocate().
+        struct ContiguousLinkNodeAllocation {
+            T object;
+            AllocationAddress address;
+        };
+
+        T* Allocate() {
+            // If a free allocation is available, reuse that location.
+            if (!mFreeQueue.empty()) {
+                AllocationAddress freeSpot = mFreeQueue.front();
+                mFreeQueue.pop();
+                return reinterpret_cast<T*>(&(*mAllocations[freeSpot.first])[freeSpot.second]);
+            }
+
+            // If our next allocation offset starts with a zero it means we haven't allocated the
+            // block yet. Allocate it.
+            if (mNextAllocation.second == 0) {
+                mAllocations.push_back(new std::vector<ContiguousLinkNodeAllocation>());
+                mAllocations[mNextAllocation.first]->reserve(mObjectsPerBlock);
+            }
+
+            // The indices for the allocation are stored on the allocation itself.
+            (*mAllocations[mNextAllocation.first])[mNextAllocation.second].address =
+                mNextAllocation;
+
+            // We can reinterpret_cast<T*> from the ContiguousLinkNodeAllocation to get a pointer to
+            // the actual object allocation on the ContiguousLinkNodeAllocation.
+            T* allocationAddress = reinterpret_cast<T*>(
+                &(*mAllocations[mNextAllocation.first])[mNextAllocation.second]);
+
+            // Increment next allocation position.
+            mNextAllocation.second++;
+            if (mNextAllocation.second == mObjectsPerBlock) {
+                mNextAllocation.first++;
+                mNextAllocation.second = 0;
+            }
+
+            return allocationAddress;
+        }
+
+        void Deallocate(T* objectPointer) {
+            ContiguousLinkNodeAllocation* allocation =
+                reinterpret_cast<ContiguousLinkNodeAllocation*>(objectPointer);
+            // Insert the AllocationAddress into mFreeQueue
+            mFreeQueue.push(allocation->address);
+        }
+
+      private:
+        uint64_t mObjectsPerBlock;
+
+        AllocationAddress mNextAllocation = {0, 0};
+
+        std::vector<std::vector<ContiguousLinkNodeAllocation>*> mAllocations;
+        std::queue<AllocationAddress> mFreeQueue;
```

Comment:
Why a queue? You only push/pop from one end.

Reply:
I `push` newly freed slots onto the back of the queue. I `pop` from the front of the queue for reuse. I use a queue because it is FIFO. I am using a queue exactly as intended as far as I can tell.

Reply:
Why do we need access on both ends?

LIFO also allows reuse and could use a std::vector, which is faster then a std::queue (which is really a std::deque under-the-hood).

Reply:
I see now. Changed to a vector.